:tada: add support of DebridLinkFr

Wald764 opened the pull request:
Hello! A year ago, I created a fork to integrate the app with DebridLink (https://debrid-link.com/). I recently noticed that you've added other providers, so with this PR, you can add a new one as well.

Wald764 commented:
https://github.com/rogerfar/rdt-client/issues/249

Cucumberrbob commented:
Thanks for the PR, I'm loving seeing more and more new debrid services and download clients getting added!

<details><summary>A slightly off topic request:</summary>
<p>

I've been working on getting the `SymlinkDownloader` to be able to figure out where the source files (files we want to create symlinks to) will end up - that is, I've been trying to simplify how `rdt-client` works out what each provider thinks is the directory structure of a given torrent.
Are you able to do a few tests with some torrents on DebridLink and let me know how they look on the webdav afterwards?

Mostly I'm interested in what a torrent with a single file looks like - AD puts these at the root of the webdav (`/mnt-root/file.mkv`), whereas RD, PM, and TB all put it in a directory (I think). If it does end up at the root of the webdav, what about a torrent containing a single directory containing a single file?
I'm also interested in deeply nested torrents - if the torrent contains a file `a/b/c/d.mkv`, AD puts this at `/mnt-root/<torrent-name>/a/b/c/d.mkv`, whereas RD puts this at `/mnt-root/<torrent-name>/d.mkv` (I think).
</p>
</details>

Cucumberrbob commented on server/RdtClient.Data/Enums/Provider.cs:
```diff
@@ -13,6 +13,9 @@ public enum Provider
     [Description("Premiumize")]
     Premiumize,
 
+    [Description("DebridLinkFr")]
+    DebridLinkFr,
+
     [Description("TorBox")]
     TorBox
```
```suggestion
    [Description("TorBox")]
    TorBox

    [Description("DebridLinkFr")]
    DebridLinkFr,
```

This way TorBox will keep `3`, and  `DebridLinkFr` will be `4`.

Wald764 replied:
done in https://github.com/rogerfar/rdt-client/pull/680/commits/d6d1813340115364e30b0f7e6b866abb277590bd

Cucumberrbob commented on server/RdtClient.Service/Services/TorrentClients/DebridLinkFrTorrentClient.cs:
```diff
@@ -0,0 +1,330 @@
+
+            if (String.IsNullOrWhiteSpace(apiKey))
+            {
+                throw new Exception("Real-Debrid API Key not set in the settings");
```
```suggestion
                throw new Exception("DebridLinkFr API Key not set in the settings");
```

Cucumberrbob replied:
@Wald764 just noticed this didn't get updated before I approved - could you make this change or make an argument why it should be `Real-Debrid` instead of `DebridLinkFr`?

Wald764 replied:
I made the change in https://github.com/rogerfar/rdt-client/pull/680/commits/f1bdc7d8b416aca96d84197b9205f9615f198e30

Cucumberrbob commented on server/RdtClient.Service/Services/TorrentClients/DebridLinkFrTorrentClient.cs:
```diff
@@ -0,0 +1,330 @@
+            Added = DateTimeOffset.FromUnixTimeSeconds(torrent.Created),
+            Files = (torrent.Files ?? new List<TorrentFile>()).Select((m, i) => new TorrentClientFile
+            {
+                Path = m.Name ?? "",
```
In what scenarios is `m.Name` null?
`Path` gets used by a few of the downloaders, so it might be worth logging  a warning message if `m.Name` is null.

Wald764 replied:
I think there is no scenarios where n.Name can be null

modified in https://github.com/rogerfar/rdt-client/pull/680/commits/d6d1813340115364e30b0f7e6b866abb277590bd

Cucumberrbob commented on server/RdtClient.Service/Services/TorrentClients/DebridLinkFrTorrentClient.cs:
```diff
@@ -0,0 +1,330 @@
+                torrent.RdFiles = JsonConvert.SerializeObject(rdTorrent.Files);
+            }
+
+            torrent.RdHost = rdTorrent.Host;
+            torrent.RdSplit = rdTorrent.Split;
+            torrent.RdProgress = rdTorrent.Progress;
+            torrent.RdAdded = rdTorrent.Added;
+            torrent.RdEnded = rdTorrent.Ended;
+            torrent.RdSpeed = rdTorrent.Speed;
+            torrent.RdSeeders = rdTorrent.Seeders;
+            torrent.RdStatusRaw = rdTorrent.Status;
```
Since a fairly recent PR, all `ITorrentClient`s set `ClientKind` on the `torrent` here: [TB](https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Service/Services/TorrentClients/TorBoxTorrentClient.cs#L233) [AD](https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Service/Services/TorrentClients/AllDebridTorrentClient.cs#L180) [PM](https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Service/Services/TorrentClients/PremiumizeTorrentClient.cs#L165) [RD](https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Service/Services/TorrentClients/RealDebridTorrentClient.cs#L298)

`DebridLinkFr` will also need to be added to the enum [`TorrentClientKind`](https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Data/Models/Data/Torrent.cs#L97) in `RdtClient.Data/Models/Data/Torrent.cs`

Wald764 replied:
debrid link added to the enum and set on the torrent 

https://github.com/rogerfar/rdt-client/pull/680/commits/d6d1813340115364e30b0f7e6b866abb277590bd

Cucumberrbob reviewed (changes requested):
Awesome PR, thanks for all the work!
A few small things could be changed imo

Wald764 commented:
> Awesome PR, thanks for all the work! A few small things could be changed imo
> 
> Also, do they prefer being called DebridLinkFr or DebridLink? It seems like they're moving away from the `.fr` domain, but I don't really know

Thank you for the review! I wasn’t entirely sure about the name either, so I removed the “fr” and renamed the provider to “DebridLink”. You can see the changes here:

https://github.com/rogerfar/rdt-client/pull/680/commits/20fbf8833a289baec465fb3d60cc4ee2d1c769ad

For your request regarding the SymlinkDownloader. I personally use the Bezzad Downloader, so unfortunately I haven't been able to test it myself.

Cucumberrbob commented on server/RdtClient.Service/Services/Torrents.cs:
```diff
@@ -23,6 +23,7 @@ public class Torrents(
     AllDebridTorrentClient allDebridTorrentClient,
     PremiumizeTorrentClient premiumizeTorrentClient,
     RealDebridTorrentClient realDebridTorrentClient,
+    DebridLinkClient DebridLinkClient,
```
is this intentionally PascalCase not camelCase like the others?

Wald764 replied:
No that was not, fixed in the last commit https://github.com/rogerfar/rdt-client/pull/680/commits/303244127bd18042b3bcb24024c9a138fac6f452

Cucumberrbob commented:
> For your request regarding the SymlinkDownloader. I personally use the Bezzad Downloader, so unfortunately I haven't been able to test it myself.

It was more a request to you as someone who uses debrid-link to look at what the file structure is like in the webdav. You won't need to actually use the SymlinkDownloader - I should be able to mostly handle this once I know the file structure in the webdav.
You can use MacOS' Finder (Connect To Server), Windows' File Explorer (Map Network Drive), or `rclone` with either Linux or MacOS.
I can probably handle this myself though - I only need a few torrents to be able to deduce the general logic behind how files will appear on the webdav, and their free tier has let me upload 2 torrents so far, it would just be slightly quicker if you did it now - so it's far from necessary.

On that note of necessity, it might be worth putting a pin in this discussion for now and making an issue along the lines of "implement `GetSymlinkPath` on `DebridLinkTorrentClient`" once this is merged - I don't want to derail your PR.

Wald764 commented:
> > For your request regarding the SymlinkDownloader. I personally use the Bezzad Downloader, so unfortunately I haven't been able to test it myself.
> 
> It was more a request to you as someone who uses debrid-link to look at what the file structure is like in the webdav. You won't need to actually use the SymlinkDownloader - I should be able to mostly handle this once I know the file structure in the webdav. You can use MacOS' Finder (Connect To Server), Windows' File Explorer (Map Network Drive), or `rclone` with either Linux or MacOS. I can probably handle this myself though - I only need a few torrents to be able to deduce the general logic behind how files will appear on the webdav, and their free tier has let me upload 2 torrents so far, it would just be slightly quicker if you did it now - so it's far from necessary.
> 
> On that note of necessity, it might be worth putting a pin in this discussion for now and making an issue along the lines of "implement `GetSymlinkPath` on `DebridLinkTorrentClient`" once this is merged - I don't want to derail your PR.

Oh, okay, the torrents are stored in a ~/Seedbox folder, so the path remains the same as the one in the torrent. I don't have a deeply nested torrent to test further; if you have one, could you please provide it?

Cucumberrbob commented:
[Torrents.zip](https://github.com/user-attachments/files/18633199/Torrents.zip)

These would be super helpful to test - thank you so much!

Cucumberrbob commented:
On [3032441](https://github.com/rogerfar/rdt-client/pull/680/commits/303244127bd18042b3bcb24024c9a138fac6f452)[^1], I'm getting errors that sound like they're HTTP [416](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416) codes with the Internal Downloader:

[^1]: [56a61ad](https://github.com/rogerfar/rdt-client/pull/680/commits/56a61ad012bb36106807131c9c5000ce8ead010e) doesn't build because I'm an idiot and didn't even vaguely check anything before pushing that commit. I've filed a PR [here](https://github.com/Wald764/rdt-client/pull/4) to fix that

<details><summary>Logs</summary>
<p>

```
[23:14:26 DBG] Updating torrent info from debrid provider
[23:14:26 DBG] Finished updating torrent info from debrid provider, next update in 10 seconds
[23:14:26 DBG] Processing 1 torrents
[23:14:26 DBG] Currently 0 queued downloads and 0 total active downloads for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:26 DBG] Processing for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:26 DBG] Selecting files for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:27 DBG] Processing 1 torrents
[23:14:27 DBG] Currently 0 queued downloads and 0 total active downloads for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:27 DBG] Processing for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:27 DBG] Creating downloads for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:27 DBG] Found 1 links for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:27 DBG] https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Processing 1 torrents
[23:14:28 DBG] Currently 1 queued downloads and 0 total active downloads for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Processing to download for download https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Unrestricting links for download https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Unrestricting link for download https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Retrieving filename for for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Marking download as started for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Setting download path to /data/downloads for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Starting download for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Instantiated new Internal Downloader for URI https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt to filePath /data/downloads/TestSingle.txt/TestSingle.txt
[23:14:28 DBG] Starting download of https://seed29.debrid.link/dl/[REDACTED]-1/TestSingle.txt, writing to path: /data/downloads/TestSingle.txt/TestSingle.txt
[23:14:28 DBG] Content size 36
[23:14:28 DBG] Starting download tasks with 1 parallel downloads
[23:14:28 DBG] Received ID bb73add0-beb6-45d7-a62e-7477c99be94c for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Started download for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID:  for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Processing for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:28 DBG] Waiting for downloads to complete. 0/1 complete (0%) for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:29 DBG] TorrentRunner Tick Start, 1 active downloads, 0 active unpacks
[23:14:29 DBG] Processing 1 torrents
[23:14:29 DBG] Currently 0 queued downloads and 1 total active downloads for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
... Repeat above 5 lines every second until...
[23:14:49 DBG] Complete
[23:14:49 DBG] TorrentRunner Tick Start, 1 active downloads, 0 active unpacks
[23:14:49 DBG] Processing 1 completed downloads
[23:14:49 DBG] Processing download for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID: bb73add0-beb6-45d7-a62e-7477c99be94c for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:49 ERR] Download reported an error: Failed to download bytes from range 0-36. Status code: RequestedRangeNotSatisfiable for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID: bb73add0-beb6-45d7-a62e-7477c99be94c for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:49 DBG] Download retry count 0/3, torrent retry count 0/1 for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID: bb73add0-beb6-45d7-a62e-7477c99be94c for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
[23:14:49 DBG] Retrying download for download TestSingle.txt. Completed: 0%, avg speed: 0bytes/s (6f585945-3a18-44ed-b163-de5a8455d84d) remoteID: bb73add0-beb6-45d7-a62e-7477c99be94c for torrent TestSingle.txt ([REDACTED] - 100 100%) (f11c6ace-9b95-4493-98f5-b2cbbbc06d7d)
```

</p>
</details>

Wald764 commented:
> On [3032441](https://github.com/rogerfar/rdt-client/pull/680/commits/303244127bd18042b3bcb24024c9a138fac6f452)[1](#user-content-fn-1-1ee8b4187d5e875de8fa834b714773f3), I'm getting errors that sound like they're HTTP [416](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416) codes with the Internal Downloader: The exact same torrent works when I retry it after changing the provider to AD
> 
> Bezzad also works perfectly.
> 

Yes, I have the same issue with the internal downloader, and I'm glad you found a solution.

Cucumberrbob commented:
PR filed [rogerfar/Downloader.NET#5](https://github.com/rogerfar/Downloader.NET/pull/5)

Cucumberrbob commented:
OK @rogerfar I think we're ready to go!
I've tested the Symlink, Internal and Bezzad downloaders - they all work.
I've tested with sonarr and radarr with single file and multi file torrents - everything works.

The only funky things are the 2 build warnings introduced:
```
0>DebridLinkTorrentClient.cs(83,24): Warning CS8601 : Possible null reference assignment.
0>DebridLinkTorrentClient.cs(157,31): Warning CS1998 : This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
```
And the nullish coalescence to the empty string in [`Map`](https://github.com/Wald764/rdt-client/blob/f1bdc7d8b416aca96d84197b9205f9615f198e30/server/RdtClient.Service/Services/TorrentClients/DebridLinkTorrentClient.cs#L69), [`AddMagnet`](https://github.com/Wald764/rdt-client/blob/f1bdc7d8b416aca96d84197b9205f9615f198e30/server/RdtClient.Service/Services/TorrentClients/DebridLinkTorrentClient.cs#L131), and [`AddFile`](https://github.com/Wald764/rdt-client/blob/f1bdc7d8b416aca96d84197b9205f9615f198e30/server/RdtClient.Service/Services/TorrentClients/DebridLinkTorrentClient.cs#L138). These will likely never be null, but this PR is potentially introducing weird behaviour if it silently converts null to the empty string.

Sculas commented:
Off-topic to this PR, but it actually made me realize that `TorrentClientKind`:

https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Data/Models/Data/Torrent.cs#L97-L103

Is actually just a duplicate of `Provider`, which I never knew existed:

https://github.com/rogerfar/rdt-client/blob/dc2ec486d7a9c5d95da47af74646588e83dc3682/server/RdtClient.Data/Enums/Provider.cs#L5-L18

We might want to look into merging these in the future or rename `TorrentClientKind` to something like `TorrentClientBehavior` which describes the file behavior instead. We can't just delete it though, since it's a database field now.

rogerfar commented:
@Sculas good fine, I'll refactor that out.
